crypto/evp/ctrl_params_translate.c

```diff
@@ -2895,11 +2895,15 @@ static int evp_pkey_ctx_setget_params_to_ctrl(EVP_PKEY_CTX *pctx,
 
 int evp_pkey_ctx_set_params_to_ctrl(EVP_PKEY_CTX *ctx, const OSSL_PARAM *params)
 {
+    if (ctx->keymgmt != NULL)
```

Comment:
I wonder if this is the correct thing to test?
It fails because no operation has been set yet I would assume.

Reply:
ctx->operation == EVP_PKEY_OP_UNDEFINED ??

Reply:
Testing against keymgmt != NULL appears to be the consistent pattern used to determine legacy/non-legacy.

Reply:
IMO @slontis is right. The key can be legacy and in such case this should proceed and not return failure.

The right check would be `ctx->operation == EVP_PKEY_OP_UNDEFINED` instead.

Reply:
The new test in `make_key_fromdata()` of `evp_extra_test.c` is exercised with many different keys (though I expect all are provided). Originally, when the operation was unset, the codepath chosen was the same as for legacy keys. And if you look at the function ultimately called: https://github.com/openssl/openssl/blob/4a1a7fe5ce088964010779e1f5a90560903ecc76/crypto/evp/ctrl_params_translate.c#L2837-L2844
You'll see that it is prominently marked as legacy-only, and explicitly supports operation being `0` (unset).
So it is rather unclear why I should be changing the handling of that case.

Reply:
Hmm... I slightly misunderstood the problem. However I believe the right fix is to modify evp_pkey_ctx_state() to return EVP_PKEY_STATE_LEGACY if `keymgmt != NULL && ctx->operation == EVP_PKEY_OP_UNDEFINED` and fix the callers of evp_pkey_ctx_state() to not call the transforms to ctrl when EVP_PKEY_STATE_UNKNOWN is detected.

Reply:
But why do we need such redundant check then?